xen/arch/arm/gic-v3-its.c

```diff
@@ -61,6 +61,76 @@ struct its_device {
     struct pending_irq *pend_irqs;      /* One struct per event */
 };
 
+/*
+ * It is unlikely that a platform implements ITSes with different quirks,
+ * so assume they all share the same.
+ */
+struct its_quirk {
+    const char *desc;
+    bool (*init)(struct host_its *hw_its);
+    uint32_t iidr;
+    uint32_t mask;
+};
+
+uint32_t its_quirk_mask;
+
+static bool gicv3_its_enable_quirk_r8a779f0(struct host_its *hw_its)
+{
+    its_quirk_mask |= HOST_ITS_WORKAROUND_R8A779F0;
+
+    return true;
+}
+
+static const struct its_quirk its_quirks[] = {
+    {
+        .desc	= "R-Car S4",
+        .iidr	= 0x0201743b,
+        .mask	= 0xffffffff,
+        .init	= gicv3_its_enable_quirk_r8a779f0,
+    },
+    {
+        /* Sentinel. */
+    }
+};
+
+static void gicv3_its_enable_quirks(struct host_its *hw_its)
+{
+    const struct its_quirk *quirks = its_quirks;
+    uint32_t iidr = readl_relaxed(hw_its->its_base + GITS_IIDR);
+
+    for ( ; quirks->desc; quirks++ )
+    {
+        if ( quirks->iidr != (quirks->mask & iidr) )
+            continue;
+        if ( quirks->init(hw_its) )
+            printk("GICv3: enabling workaround for ITS: %s\n", quirks->desc);
+    }
+}
+
+uint64_t gicv3_its_get_cacheability(void)
+{
+    if ( its_quirk_mask & HOST_ITS_WORKAROUND_R8A779F0 )
+        return GIC_BASER_CACHE_nC;
+
+    return GIC_BASER_CACHE_RaWaWb;
+}
+
+uint64_t gicv3_its_get_shareability(void)
+{
+    if ( its_quirk_mask & HOST_ITS_WORKAROUND_R8A779F0 )
+        return GIC_BASER_NonShareable;
+
+    return GIC_BASER_InnerShareable;
+}
+
+unsigned int gicv3_its_get_memflags(void)
+{
+    if ( its_quirk_mask & HOST_ITS_WORKAROUND_R8A779F0 )
```

Comment:
I'm not sure that we need one `HOST_ITS_WORKAROUND_R8A779F0` flag to rule them all.

We need to check: are there other ITS implementation that support only 32 bit addresses? Maybe we should have flags like `HOST_ITS_WORKAROUND_32BIT_ADDRESSES` and `HOST_ITS_WORKAROUND_NC_NS` at least.

Reply:
agree